File: Sorting/18-almost-sorted.php
<?php

$handle = fopen("php://stdin", "r");
fscanf($handle, "%d", $n);

$origin = array_map('intval', explode(' ', trim(fgets($handle))));
$sorted = $origin;
sort($sorted);

$diff = [];
for ($i=0; $i < $n; $i++) {
    if ($origin[$i] != $sorted[$i]) {
        $diff[] = $i;
    }
}

$count = count($diff);

//sorted
if ($count == 0) {
    echo "yes\n";
} else if ($count == 2 && // swap
    $sorted[$diff[0]] == $origin[$diff[1]] &&
    $sorted[$diff[1]] == $origin[$diff[0]]) {
    printf("yes\nswap %d %d\n", $diff[0] + 1, $diff[1] + 1);
} else if ($count > 2) {
    $reverse = false;
    for ($i=0; $i < $count; $i++) {
        if ($sorted[$diff[$i]] == $origin[$diff[$count - 1 - $i]]) {
            $reverse = true;
        } else {
            $reverse = false;
            break;
        }
    }
    if ($reverse) {
        printf("yes\nreverse %d %d\n", $diff[0] + 1, $diff[$count - 1] + 1);
    } else {
        echo "no\n";
    }
}

File: Algorithm/Sorting/18-almost-sorted.cpp
#include <iostream>
#include <vector>
#include <algorithm>

typedef long long LL;

using namespace std;

int main(int argc, char const *argv[])
{
    LL n;
    vector <LL> origin, sorted, diff;
    cin >> n;
    for (LL i = 0; i < n; ++i) {
        LL x;
        cin >> x;
        origin.push_back(x);
        sorted.push_back(x);
    }
    sort(sorted.begin(), sorted.end());

    for (LL i = 0; i < n; ++i) {
        if (origin[i] != sorted[i]) {
            diff.push_back(i);
        }
    }

    LL count = diff.size();

    if (count == 0) {
        cout << "yes" << endl;
        return 0;
    }

    if (count == 2 &&
        sorted[diff[0]] == origin[diff[1]] &&
        sorted[diff[1]] == origin[diff[0]]) {
        cout << "yes" << endl << "swap " << diff[0] + 1 << " " << diff.back() + 1;
        return 0;
    }

    bool reverse = false;
    for (LL i = 0; i < count; i ++) {
        if (origin[diff[i]] == sorted[diff[count - i - 1]]) {
            reverse = true;
        } else {
            reverse = false;
            break;
        }
    }

    if (reverse) {
        cout << "yes" << endl << "reverse " << diff[0] + 1 << " " << diff.back() + 1;
    } else {
        cout << "no" << endl;
    }

    return 0;
}
